source/common/websocket/codec.cc

```diff
@@ -0,0 +1,212 @@
+  }
+  // Set masking key
+  if (frame.masking_key_) {
+    pushScalarToByteVector(htobe32(frame.masking_key_.value()), output);
+  }
+  return output;
+}
+
+void Decoder::frameMaskFlag(uint8_t mask_and_length) {
+  num_remaining_masking_key_bytes_ = mask_and_length & 0x80 ? kMaskingKeyLength : 0;
+  length_ = mask_and_length & 0x7F;
+}
+
+void Decoder::frameDataStart() {
+  frame_.payload_length_ = length_;
+  if (length_ == 0) {
+    state_ = State::FrameFinished;
+  } else {
+    frame_.payload_ = std::make_unique<Buffer::OwnedImpl>();
+    state_ = State::FramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd(uint64_t& bytes_consumed_by_frame, Buffer::Instance& input,
+                           absl::optional<std::vector<Frame>>& output) {
+  output->push_back(std::move(frame_));
+  resetDecoder();
+  input.drain(bytes_consumed_by_frame);
+  bytes_consumed_by_frame = 0;
+}
+
+void Decoder::resetDecoder() {
+  frame_ = {false, 0, absl::nullopt, 0, nullptr};
+  state_ = State::FrameHeaderFlagsAndOpcode;
+  length_ = 0;
+  num_remaining_extended_length_bytes_ = 0;
+  num_remaining_masking_key_bytes_ = 0;
+}
+
+uint8_t Decoder::doDecodeFlagsAndOpcode(absl::Span<const uint8_t>& data) {
+  // Validate opcode (last 4 bits)
+  uint8_t opcode = data.front() & 0x0f;
+  if (std::find(kFrameOpcodes.begin(), kFrameOpcodes.end(), opcode) == kFrameOpcodes.end()) {
+    ENVOY_LOG(debug, "Failed to decode websocket frame with invalid opcode: {}", opcode);
+    return 0;
+  }
+  frame_.opcode_ = opcode;
+  frame_.final_fragment_ = data.front() & 0x80;
+  state_ = State::FrameHeaderMaskFlagAndLength;
+  return 1;
+}
+
+uint8_t Decoder::doDecodeMaskFlagAndLength(absl::Span<const uint8_t>& data) {
+  frameMaskFlag(data.front());
+  if (length_ == 0x7e) {
+    num_remaining_extended_length_bytes_ = kPayloadLength16Bit;
+    length_ = 0;
+    state_ = State::FrameHeaderExtendedLength16Bit;
+  } else if (length_ == 0x7f) {
+    num_remaining_extended_length_bytes_ = kPayloadLength64Bit;
+    length_ = 0;
+    state_ = State::FrameHeaderExtendedLength64Bit;
+  } else if (num_remaining_masking_key_bytes_ > 0) {
+    state_ = State::FrameHeaderMaskingKey;
+  } else {
+    frameDataStart();
+  }
+  return 1;
+}
+
+uint8_t Decoder::doDecodeExtendedLength(absl::Span<const uint8_t>& data) {
+  uint64_t bytes_to_decode = data.length() <= num_remaining_extended_length_bytes_
+                                 ? data.length()
+                                 : num_remaining_extended_length_bytes_;
+  uint8_t* destination = reinterpret_cast<uint8_t*>(&length_) +
+                         (state_ == State::FrameHeaderExtendedLength16Bit ? kPayloadLength16Bit
+                                                                          : kPayloadLength64Bit) -
+                         num_remaining_extended_length_bytes_;
```

Comment:
Maybe worth an assertion here ensure destination over the boundary like (destination - length_) < 8